src/api/clients/emailNotifications.ts

```diff
@@ -0,0 +1,49 @@
+import config from 'app/config';
+import { EmailValidationError } from 'app/errors';
+
+import createHttpClient from '../client';
+import {
+  SubscribePayload,
+  UnsubscribePayload,
+  AuthenticatePayload,
+  ModifyPayload,
+  CheckSubscriptionPayload,
+  VerifyEmailPayload,
+  VerifyEmailResponse,
+} from '../types';
+
+const httpClient = createHttpClient(config.emailNotificationsApi);
+
+export const subscribeEmail = async (data: SubscribePayload) => {
+  try {
+    const response = await httpClient.post(`/subscribe/`, data);
+
+    return response.data;
+  } catch (error) {
```

Comment:
I'm not a fan of this concept of wrapping requests manually. We should create something generic for all requests. My proposal:

`const httpClient = createHttpClient(config.emailNotificationsApi, errorHandler)`
Inside the httpClient
```
const onError = (error: AxiosError) => {
    if (error.response) {
      logger.error('http', `<-- ${error.response.status} ${error.config.baseURL}${error.config.url}`);

      if (error.response.status === 400) {
        throw new HttpBadRequestError(error.response.data.msg);
      }

      if (error.response.status === 406) {
        throw new HttpNotAcceptableError(error.response.data.msg);
      }

      if (error.response.status === 500) {
        throw new HttpTechnicalError(error.response.data.msg);
      }
    }
    errorHandler(error)
  };
  ```
  error handler 
  ```
  const errorHandler = (error) =>{
  if (error.message.includes('Enter a valid email address')) {
      throw new EmailValidationError();
    }
    ....
    }
 ```

Reply:
Hm... I'm wondering if your approach won't affect readability. Let's say, you want to find out what's the behaviour of the `/verify_email/` endpoint, what are the negative server responses  - following the generic error handling approach we lose this information. We can only guess in the case of the invalid email address the response body will include `msg` with the string "Enter a valid email address" but you don't have this information explicitly wrote in the code.

Also, with this approach, we must trust that backend works consistently and, for instance, it always responds with the same message if the given email address is invalid. Of course, it should work consistently but I'm not convinced if we should design our code based on faith. Right now, we integrate with the backend developed in-house but assuming that in the future it might change then we won't have any influence on how the service is developed.

I think we can mix two approaches somehow

Reply:
in the code now we have error.message.includes('Enter a valid email address') so we also trust here that the message will be like this, there is no difference in that matter, the logic is just moved to the global handler.
For the errors documentation per endpoint, we have swagger, all error should be documented there. What we really care about here is mapping error messages to proper errors with the translated messages, so the sagas can catch it.

A different solution would be not to catch EmailValidationError, let throw the message and translate the error message in the selector. I would say I like even more this solution because we don't need instances of error from the backend, what we need is to translate those messages. What I would do in onError function for the httpClient would be just mapping error.response.data.msg to error.message so we will have the same error format for every error
<img width="1007" alt="Zrzut ekranu 2021-02-3 o 15 46 01" src="https://user-images.githubusercontent.com/67273035/106763273-f20ae700-6636-11eb-8553-78f2485f356c.png">

Reply:
Late to the party, but I would go with generic approach as well. Also would it be possilble to not rely on strings comparison when checking errors? Checking string inclusion in one place, comparing strings in another - so many places for errors - both in code and in contract with BE. Maybe it would be better to introduce additional abstraction layer, like error codes. 
1, 2, 3, 4, 5, 6 .... mapped to messages. One map in backend, one in frontend, documented in swagger. 
`{ 1:  "Invalid email",  2: "Server unresponsive"  }` etc. 
This way, we can check error code (not much space for error here) and take error from map of errors, that already would have translated values: 
`{ 1: i18n.errors.invalidEmail, 2: i18n.error.serverUnresponsive }`. 
We can store this map in single file, so it s easy to fetch it `errorMap[error.response.data.code]` and then also easy to compare it when we lose consistency (easier to compare maps like this, then string that are scattered in multiple files)

Reply:
yeah agreed, I think we should have some kind of mapper for common errors like server not responding etc. To sum up, 
 I would do something like this:
API client
```
const onError = (error: AxiosError) => {
    if (error.response) {
      logger.error('http', `<-- ${error.response.status} ${error.config.baseURL}${error.config.url}`);

      throw new Error(error.response.data.msg) // mapping the error so they all have some format
    }
  };
  ```
  API notifications
  ```
  import config from 'app/config';
import { EmailValidationError } from 'app/errors';

import createHttpClient from '../client';
import {
  SubscribePayload,
  UnsubscribePayload,
  AuthenticatePayload,
  ModifyPayload,
  CheckSubscriptionPayload,
  VerifyEmailPayload,
  VerifyEmailResponse,
} from '../types';

const httpClient = createHttpClient(config.emailNotificationsApi);

export const subscribeEmail = (data: SubscribePayload) => httpClient.post(`/subscribe/`, data);

export const unsubscribeEmail = (data: UnsubscribePayload) => httpClient.post(`/unsubscribe/`, data);

export const authenticateEmail = (data: AuthenticatePayload) => httpClient.post(`/authenticate/`, data);

export const modifyEmail = (data: ModifyPayload) => httpClient.put(`/modify/`, data);

export const checkSubscriptionEmail = (data: CheckSubscriptionPayload) => httpClient.post(`/check_subscription/`, data);

export const verifyEmail = (data: VerifyEmailPayload): Promise<VerifyEmailResponse> => httpClient.post(`/verify_email/`, data);
```
sagas notification
```
export function* verifyNotificationEmailSaga(action: VerifyNotificationEmailAction | unknown) {
  const { meta, payload } = action as VerifyNotificationEmailAction;
  const email = payload.email;

  try {
    const verifyCode = yield call(verifyEmail, { email });
    const decryptedCode = yield decryptCode(email, verifyCode.pin);

    yield put(verifyNotificationEmailSuccess(decryptedCode));

    if (meta?.onSuccess) {
      meta.onSuccess();
    }
  } catch (error) {
     yield put(verifyNotificationEmailFailure(error.message));

    if (meta?.onFailure) {
      meta.onFailure();
    }
  }
}
```
selectors notification
```
export const readableError = createSelector(notificationError, failedTries, (err, failNo) => {
 // mapper error function for common errors here
 
 // some notification specific errors here
  if (err === messages.invalidEmail) {
    return i18n.notifications.invalidAddressError;
  }

  if (err === messages.wrongPIN && failNo < CONST.emailCodeErrorMax) {
    return i18n.formatString(i18n.notifications.codeError, { attemptsLeft: CONST.emailCodeErrorMax - failNo });
  }
  return err;
});
```

Reply:
Okay, let's do that this way then 🙂